### PR TITLE
Feat/475 add manual stepping option for turing machines in testing lab

### PR DESCRIPTION
**Pull request by s3782534**
The core functionality is complete, and some obvious bugs have already been fixed. Would like someone else to look over it to make sure it's good enough.

### CODE REVIEW

File: frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx [hunk 1]
```diff
@@ -248,9 +310,61 @@ const TestingLab = () => {
...
-  }, [traceInput, simulationResult, statePrefix, traceIdx, getStateName])
+    return `${transitionsWithRejected.join('\n')}${(traceIdx === lastTraceIdx || (enableManualStepping && currentManualSuccessors.length === 0)) ? '\n\n' + (accepted ? 'ACCEPTED' : 'REJECTED') : ''}`
+  }, [traceInput, simulationResult, statePrefix, traceIdx, getStateName, enableManualStepping])
 
+  // Creates problem and sets it, should run every time trace/graph changes while enableManualStepping is enabled
+  useEffect(() => {
+    if (enableManualStepping) {
+      const _problem = buildProblem(graph, traceInput)
+      if (_problem != null) {
+        console.log('currentManualNode reset')
```
Comment (s3897720) [inline frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx:321]: You may want to comment out the console.log() calls to make it a bit cleaner?
```suggestion
    
```

File: frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx [hunk 2]
```diff
@@ -55,6 +68,54 @@ const TestingLab = () => {
...
+          assertType<Node<TMState>>(node)
+          const trace = generateTraceTM(node)
+          result = {
+            accepted: node.state.isFinal,
+            tape: node.state.tape,
+            trace
+          }
+        }
+    }
+    console.log(['result is:', result, enableManualStepping, node, currentManualNode])
```
Comment (s3897720) [inline frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx:115]: ditto
```suggestion
    
```

File: frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx [hunk 3]
```diff
@@ -248,9 +310,61 @@ const TestingLab = () => {
...
+  useEffect(() => {
+    if (enableManualStepping) {
+      const _problem = buildProblem(graph, traceInput)
+      if (_problem != null) {
+        console.log('currentManualNode reset')
+        setProblem(_problem)
+        setCurrentManualNode(_problem.initial)
+        setTraceIdx(0)
+      } else {
+        console.log('Problem not created properly!')
```
Comment (s3897720) [inline frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx:326]: Ditto
```suggestion
        
```

File: frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx [hunk 4]
```diff
@@ -289,6 +403,17 @@ const TestingLab = () => {
...
   const automataIsInvalid = noInitialState || noFinalState || !pathToFinal
 
+  // Display the valid transitions that could be manually chosen
+  const buttonsArray:JSX.Element[] = []
+  currentManualSuccessors.forEach(t => {
+    // Add transition to the form of <Button>{'a,X;R: q0 -> q1'}</Button>
+    buttonsArray.push(<Button onClick={() => {
+      setCurrentManualNode(t)
+      setTraceIdx(traceIdx + 1)
+      console.log(['New node set!', currentManualNode, currentManualSuccessors])
```
Comment (s3897720) [inline frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx:413]: Ditto
```suggestion
     
```

**Review by s3897720 — APPROVED** [review]
Let's see if I can break GitHub by approving this while having open discussions preventing this from being merged

File: frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx (hunk 1 above)
Comment (s3782534) [reply frontend/src/components/Sidepanel/Panels/TestingLab/TestingLab.tsx:321]: Yeah, forgot to remove those

**Comment by s3782534** [comment]
> Let's see if I can break GitHub by approving this while having open discussions preventing this from being merged

The approval went through but the merge was blocked by the open conversations

**Pull request merged**